src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,1840 @@
+{
+  symbol_exprt b = string_exprt::fresh_symbol(prefix,bool_typet());
+  boolean_symbols.push_back(b);
+  return b;
+}
+
+
+string_exprt string_constraint_generatort::string_of_expr(const exprt & unrefined_string)
+{
+  string_exprt s;
+
+  check_char_type(unrefined_string);
+
+  if(unrefined_string.id() == ID_function_application)
+    {
+      exprt res = function_application(to_function_application_expr(unrefined_string));
+      assert(res.type() == refined_string_typet(get_char_type()));
+      s = to_string_expr(res);
+    }
+  else if(unrefined_string.id()==ID_symbol)
+    s = get_string_of_symbol(to_symbol_expr(unrefined_string));
+  else if(unrefined_string.id()==ID_address_of)
+    {
+      assert(unrefined_string.op0().id()==ID_symbol);
+      s = get_string_of_symbol(to_symbol_expr(unrefined_string.op0()));
+    }
+  else if(unrefined_string.id()==ID_if)
+    s = string_if(to_if_expr(unrefined_string));
+  else if(unrefined_string.id()==ID_nondet_symbol || unrefined_string.id()==ID_struct) {
+    // We ignore non deterministic symbols and struct
+  }
+  else
+    {
+      throw ("string_exprt of:\n" + unrefined_string.pretty()
+	     + "\nwhich is not a function application, a symbol or an if expression");
+    }
+
+  axioms.emplace_back(s.longer(refined_string_typet::index_zero()));
+  return s;
+}
+
+
+
+string_exprt string_constraint_generatort::string_if(const if_exprt &expr)
+{
+  string_exprt res(get_char_type());
+  assert(refined_string_typet::is_unrefined_string_type(expr.true_case().type()));
+  string_exprt t = string_of_expr(expr.true_case());
+  assert(refined_string_typet::is_unrefined_string_type(expr.false_case().type()));
+  string_exprt f = string_of_expr(expr.false_case());
+
+  axioms.emplace_back(expr.cond(),res.same_length(t));
+  symbol_exprt qvar = fresh_univ_index("QA_string_if_true");
+  axioms.push_back(string_constraintt(expr.cond(),equal_exprt(res[qvar],t[qvar])
+				      ).forall(qvar,t.length()));
+
+  axioms.emplace_back(not_exprt(expr.cond()),res.same_length(f));
+  symbol_exprt qvar2 = fresh_univ_index("QA_string_if_false");
+  axioms.push_back(string_constraintt(not_exprt(expr.cond()),
+				      equal_exprt(res[qvar2],f[qvar2])
+				      ).forall(qvar2,f.length()));
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::get_string_of_symbol(const symbol_exprt & sym)
+{
+  irep_idt id = sym.get_identifier();
+  std::map<irep_idt, string_exprt>::iterator f = symbol_to_string.find(id);
+  if(f != symbol_to_string.end())
+    return f->second;
+
+  symbol_to_string[id]= string_exprt(get_char_type());
+  return symbol_to_string[id];
+}
+
+string_exprt string_constraint_generatort::string_of_symbol(const symbol_exprt & sym)
+{ return get_string_of_symbol(sym); }
+
+
+exprt string_constraint_generatort::function_application
+(const function_application_exprt & expr)
+{
+  const exprt &name = expr.function();
+  assert(name.id() == ID_symbol);
+
```

Comment:
In java, during the verification condition generation the ids gets some suffix added (for instance we get `__CPROVER_uninterpreted_string_of_int_func#0`) so I needed starts_with here. 
If there is a way to avoid getting this suffix then it would be a better to use equality.